File: src/components/pipeline/NotebookService.ts

import { toast } from "@/hooks/use-toast";
import { generateNotebookCode } from "./SparkCodeGenerator";
import { Node } from "./types";

export const deployToDatabricks = async (
  node: Node, 
  workspaceUrl: string,
  token: string,
): Promise<void> => {
  if (!workspaceUrl || !token || !node.data.fileFormat || !node.data.filePath) {
    toast({
      title: "Deployment failed",
      description: "Missing required information for deployment",
      variant: "destructive",
    });
    return;
  }

  try {
    // Generate notebook content based on the node data
    const notebookContent = generateNotebookCode({
      fileFormat: node.data.fileFormat,
      filePath: node.data.filePath,
      hasHeader: node.data.options?.hasHeader as boolean || false,
      inferSchema: node.data.options?.inferSchema as boolean || false,
      multiLine: node.data.options?.multiLine as boolean || false,
      cloudProvider: node.data.cloudProvider || 'local' // Add the cloudProvider property
    });
    
    // Create a notebook name based on the source type
    const notebookName = `Pipeline_${node.data.fileFormat.toUpperCase()}_Source_${Date.now()}`;
    
    // Create the notebook in Databricks workspace
    const response = await fetch(`${workspaceUrl}/api/2.0/workspace/import`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        path: `/Users/${notebookName}`,
        format: 'SOURCE',
        language: 'PYTHON',
        content: Buffer.from(notebookContent).toString('base64'),
        overwrite: true
      }),
    });

    if (response.ok) {
      toast({
        title: "Notebook deployed successfully",
        description: `Created notebook: ${notebookName}`,
      });
    } else {
      const errorData = await response.json();
      throw new Error(errorData.message || 'Failed to deploy notebook');
    }
  } catch (error) {
    console.error('Error deploying to Databricks:', error);
    toast({
      title: "Deployment failed",
      description: error instanceof Error ? error.message : 'Unknown error',
      variant: "destructive",
    });
  }
};

File: python/services/notebook_service.py
import base64
import requests
import time

def deploy_to_databricks(node, workspace_url, token):
    """Deploy a notebook to Databricks workspace"""
    if not workspace_url or not token or not node.get('data', {}).get('fileFormat') or not node.get('data', {}).get('filePath'):
        return {"success": False, "message": "Missing required information for deployment"}
    
    try:
        # Generate notebook content based on the node data
        from services.spark_code_generator import generate_notebook_code
        
        notebook_content = generate_notebook_code({
            "fileFormat": node.get('data', {}).get('fileFormat'),
            "filePath": node.get('data', {}).get('filePath'),
            "hasHeader": node.get('data', {}).get('options', {}).get('hasHeader', False),
            "inferSchema": node.get('data', {}).get('options', {}).get('inferSchema', False),
            "multiLine": node.get('data', {}).get('options', {}).get('multiLine', False),
            "cloudProvider": node.get('data', {}).get('cloudProvider', 'local')
        })
        
        # Create a notebook name based on the source type
        notebook_name = f"Pipeline_{node.get('data', {}).get('fileFormat', '').upper()}_Source_{int(time.time())}"
        
        # Create the notebook in Databricks workspace
        response = requests.post(
            f"{workspace_url}/api/2.0/workspace/import",
            headers={
                'Authorization': f'Bearer {token}',
                'Content-Type': 'application/json',
            },
            json={
                'path': f'/Users/{notebook_name}',
                'format': 'SOURCE',
                'language': 'PYTHON',
                'content': base64.b64encode(notebook_content.encode()).decode(),
                'overwrite': True
            }
        )
        
        if response.status_code == 200:
            return {"success": True, "message": f"Notebook {notebook_name} created successfully"}
        else:
            return {"success": False, "message": f"Failed to create notebook: {response.text}"}
            
    except Exception as e:
        return {"success": False, "message": f"Error: {str(e)}"}